src/hooks/useAsync.ts

```diff
@@ -46,11 +50,39 @@ const useAsync = <Result, Error>(
   }, [asyncFn]);
 
   return result;
-
-  // Loading is false but result hasn't been set;
 };
 
-export default useAsync;
+export type UseAsyncOnDemandReturnType<Result, Error = unknown> = [
+  () => void,
+  UseAsyncReturnType<Result, Error>
+];
+
+export const useAsyncOnDemand = <Args extends unknown[], Result, Error>(
+  asyncFn: (...args: Args) => Promise<Result>
+): UseAsyncOnDemandReturnType<Result, Error> => {
+  const [result, setResult] = useState<UseAsyncReturnType<Result, Error>>({
+    loading: false,
+    result: null,
+    error: null,
+  });
+
+  const funcOnDemand = useCallback(
+    (...args: Args) => {
+      setResult({ result: null, error: null, loading: true });
+
+      asyncFn(...args)
+        .then((result) => {
+          setResult({ result, error: null, loading: false });
+        })
+        .catch((error) => {
+          setResult({ result: null, error, loading: false });
+        });
```

Comment:
I'd use `await` here to wait for the result of `asyncFn`, and then I'd return those results from the callback. That would let you do either:
```ts
const [fn, result] = useAsyncOnDemand(...);
fn();
return <div>{result}</div>
```
OR
```
const [fn] = useAsyncOnDemand(...);
const onSubmit = useCallback(() => {
  const result = await fn();
  toast(`Successfully saved ${result.displayName}`)
})